js/modules/k6/experimental/streams/tests/readable-streams/bad-strategies.js

```diff
@@ -0,0 +1,198 @@
+// Original source file: https://github.com/web-platform-tests/wpt/blob/f0d6f241e3a16ddd6e0a3ebcd9cb814b408bc5a6/streams/readable-streams/bad-strategies.any.js
```

Comment:
I prefer to have this be downloaded similar to how `tc39` tests are that way I won't need to wonder if this file is edited or not. And also make it more obvious that we get it from another repo.

THe licensing around just copying some files in your repo are bit iffy and likely we will need to put some LICENSE file in this folder otherwise.


https://github.com/grafana/k6/blob/master/js/tc39/checkout.sh

https://github.com/grafana/k6/blob/e6e1016137667c3089e6f8b10b62e4bbefe1f2ea/js/tc39/tc39_test.go#L753-L759

Reply:
I'd love to be able to do so, as it would make the whole process much easier, but as I stated in the aforementioned document, WPTs aren't applicable as-is, as they have parts that aren't fully compatible with k6/goja. Not to mention that we don't fully support all the test suites/tests.

If copy-pasting is not a solution, we'll need to find another one, but I feel like we most likely will need to have our own version of that thing. Any suggestion? Adding a LICENSE and/or an explanation file sounds good to me, but I'm not a legal expert either.

Reply:
> s they have parts that aren't fully compatible with k6/goja

which parts are those?

> Not to mention that we don't fully support all the test suites/tests.

You can always run *some* of the tests. the tc39 test also do not run all the tests. Or you can (as our version of the tc39 tests) - run the tests but expect that they will error out. The later also helps with showcasing improvements and makes it a lot easier when adding functionality as you don't need   to also go hunt down a bunch of tests that you need to enable. Arguably you still need to go find out if some of the tests that should pass are ... well passing :)

Reply:
> which parts are those?

On the one hand, there are JS constructions that I think we don't support (yet?), like:
- [`for await...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of)
- [`async function*`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function*)

On the other hand, Web Platform Tests are mainly intended to be executed in a browser, and the ~5k lines of "basic" harness plus the surrounding code assumes that, most of the times. Of course, I can try to dig as deep as possible to that harness and to the Python code (which mainly glues a browser instance with the code) used to run Web Platform Tests in their standard way, and try to see if I can find a way to run them without modifying them at all, but overriding some parts. However, I'm not sure it really worths the effort, especially considering that we're only using that for ReadableStream tests, and that apparently we don't plan to expose such functionality as something available for extensions or more widely/generally).

> You can always run some of the tests

I'm not super convinced it's that easy to just run certain tests, especially considering that we don't want to modify the `wpt` repository contents at all, and considering what I mentioned above. In the worst case I guess we could just disable the test suites that we cannot fully support, and just enable them when modifications are required, or some workaround like that 😓

Reply:
For further reference, you can see that from [`testharness.js API`](https://web-platform-tests.org/writing-tests/testharness-api.html), it states:
 _"The test harness script can be used from HTML or SVG documents and workers."_
 
 Also, for instance:
_"By default tests must be created before the load event fires."_
referring to the [`window.load` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event), which I'm not sure it makes lot of sense in the context of k6/goja.

and more, and more...


...which doesn't necessarily mean it's impossible to run them as they are with some tooling in place, but I guess the purpose of this issue is to bring `ReadableStream` support to k6, not building our own custom runner of WPTs based on goja, which sounds like it could be a completely separated project on its own.

So, unless I'm missing something and going the wrong way, I'd prefer to back to the beginning and re-evaluate how feasible would be to merge this as-is now, and see what would be required (adding a README/LICENSE? what else?).

> unless I'm missing something and going the wrong way 
if that's the case, I apologize in advance 🙏🏻  and would like to ask you, please, to shed some light, as I've been diving through swampy waters so far 😓 

Thanks! 🙏🏻

Reply:
> On the one hand, there are JS constructions that I think we don't support (yet?), like:

Both of those are https://github.com/dop251/goja/issues/555 essentially, so yeah - not supported and no idea when they will be,but likely after ESM (given that babel will block their usage either way). 

I don't really feel the need for a dedicated repo as the usage of them for extension have shown IME that they add more overhead than actual benefit. They are used because this is how we started - separate repo for an extension that we then wanted in core.

My whole take here is - I would prefer to get the test and harness from the original form in some repeatable way. For tc39 tests that is doable through cloning the repo. If here that will mean that we can get 90% of the test we run by just cloning and the rest 10% will need to apply "cool.patch" to them, I think that is also fine. Percentages are examples. 

In another month or two (or worse a year) someone will try to add something else and will spend days if not weeks figuring out how to get this tests and what was changed to make them work.

If that is too complicated to automate I would argue it also needs a considerable *in repo* explanation on how to do it. 

I managed to automate the test retrieval and patching in apparently an hour. Of which 40 minutes was me trying to figure what was changed in the harness, and remembering that the harness isn't in the module lost me a bunch of time trying to find out `step_timeout`.

https://gist.github.com/mstoykov/284e8ef15a07ed82521b52d389c2c354

You need to run `clone-tests.sh` (which should be automated in the CI) and everything  else seems to work. 

I really would prefer if the harness was also not modified or at least was *less* modified. But I don't really want to do that as well.

Again this is similar to how we do stuff for tc39 tests. There it just so happens that you come at the time most of the tests and the harness *work* in k6. This didn't used to be the case, and re-enabling stuff was hard even when we didn't rewrite and copy-paste with changes source code.

Reply:
> I managed to automate the test retrieval and patching in apparently an hour.

That's not the complex part, but the harness, which is the what's mostly under the assumption of being executed in a browser. 

> My whole take here is - I would prefer to get the test and harness from the original form in some repeatable way. For tc39 tests that is doable through cloning the repo. If here that will mean that we can get 90% of the test we run by just cloning and the rest 10% will need to apply "cool.patch" to them, I think that is also fine. Percentages are examples.

But yeah, if having a `local.patch` for that, despite the conflicts it may cause in the future when retrieving the remote source, is acceptable, then I guess I can get something that works, much faster.

<br/>

Let me work on a proposal and see how acceptable would be what I get.

Reply:
>  that, despite the conflicts it may cause in the future when retrieving the remote source

I would prefer to have to figure out how to update  a patch once a year when we update the tests. Than having to figure out what changed across all the tests and harnessess in this time and what were our changes.

As you saw we have pinned both the tc39 tests and we will be pinning this to a version. 

> Let me work on a proposal and see how acceptable would be what I get.

:+1: Again I prefer if we changed their sources *less* and try to make them work as much as possible. Deleting parts of them as we do not support syntax or something is way more acceptable and easy to figure out in a year, than it is to rewrite the whole thing.

Reply:
Resolving this as well as the currently proposed solution in the PR works great IMO